crates/matrix-sdk/src/room/joined.rs

```diff
@@ -686,12 +686,32 @@ impl Joined {
             #[cfg(feature = "image-proc")]
             let data_slot;
             #[cfg(feature = "image-proc")]
-            let thumbnail = if config.generate_thumbnail {
-                match generate_image_thumbnail(
-                    content_type,
-                    Cursor::new(&data),
-                    config.thumbnail_size,
-                ) {
+            let (data, thumbnail) = if config.generate_thumbnail {
+                let content_type = content_type.clone();
+
+                #[cfg(not(target_arch = "wasm32"))]
+                let (data, res) = tokio::task::spawn_blocking(move || {
+                    let res = generate_image_thumbnail(
+                        &content_type,
+                        Cursor::new(&data),
+                        config.thumbnail_size,
+                    );
```

Comment:
Instead of copy-pasting this, can you define a closure like

```rust
let make_thumbnail = move |data| { ... };

#[cfg(not(target_arch = "wasm32"))]
let (data, res) = tokio::task::spawn_blocking(make_thumbnail(data)).await.expect("Task join error");

#[cfg(target_arch = "wasm32")]
let (data, res) = make_thumbnail(data);
```